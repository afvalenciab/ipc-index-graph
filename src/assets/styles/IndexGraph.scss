
.index--graph {
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  margin: 0px 50px;
  padding: 15px 10px;
  background: #FFF;
  border-radius: 10px;

  .graph--header {
    display: flex;

    .graph--date {
      padding: 5px 10px;
      font-weight: 900;
      letter-spacing: 1.5px;
      font-size: 18px;
    }

    .graph--current_value {
      padding: 5px 10px;
      
      .current_text {
        font-weight: 700;
      }

      .current_per {
        padding-left: 5px;
      }
      .current_per.down {
        color: #FD5F5D;
      }
      .current_per.up {
        color: rgb(6, 177, 80);
      }
    }

    .graph--options {
      display: flex;

      figure {
        margin: 0px 8px;
        background: rgb(210, 212, 216);
        padding: 5px;
        border-radius: 5px;
        cursor: pointer;
        will-change: transform;
        opacity: .6;

        img{
          vertical-align: top;
          width: 25px;
        }
      }

      figure:active {
        transform: scale(.95);
      }

      figure:hover {
        opacity: 1;
      }

      figure.isActive {
        opacity: 1;
      }

      .graph--candle_options {
        display: flex;

        .candle__min {
          margin: 3px;
          border-radius: 50%;
          background: rgb(237, 236, 236);
          width: 25px;
          height: 25px;
          font-size: 10px;
          text-align: center;
          line-height: 26px;
          font-weight: 300;
          cursor: pointer;
          box-sizing: border-box;
        }

        .candle__min.isActive {
          background: rgba(0,0,0,.8);
          border: none;
          color: #fff;
          font-weight: 700;
        }
      }
    }
  }

  .graph--timeseries{
    height: 400px;
  }
}