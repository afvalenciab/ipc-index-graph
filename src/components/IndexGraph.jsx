import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import Chart from 'react-apexcharts';
import moment from 'moment';
import {getConfigChart, initConfigChart} from '../utils/getConfigChart';
import {getConfigChartCandle, initConfigChartCandle} from '../utils/getConfigChartCandle';
import areaIcon from '../assets/static/stats-dots.svg';
import candleIcon from '../assets/static/candlestick-chart.svg';
import '../assets/styles/IndexGraph.scss';

const IndexGraph = (props) => {
  const { historicalIpcList, currentIpcValue, maximumIpcValue, minimumIpcValue } = props;
  const symbolPercentage = currentIpcValue.Porcentaje >= 0;
  moment.locale('es');

  const [timeSeriesChart, setTimeSeriesChart] = useState({
    area: initConfigChart,
    candle: initConfigChartCandle
  });

  const [showGraph, setShowGraph] = useState({
    area: true,
    candle: false
  });

  const [timeGraphCandle, setTimeGraphCandle] = useState({
    five: false,
    fifteen: true,
    thirty: false,
    hour: false
  });

  const handleClickArea = () => {
    setShowGraph({
      area: true,
      candle: false
    });
  };

  const handleCliclCandle = () => {
    setShowGraph({
      area: false,
      candle: true
    });
  };

  const handleChangeTimeCandle = (event) => {
    setShowGraph({
      area: false,
      candle: false
    });

    const time = Number(event.target.id);
    setTimeGraphCandle({
      five: time === 5? true : false,
      fifteen: time === 15? true : false,
      thirty: time === 30? true : false,
      hour: time === 60? true : false
    });

    setTimeSeriesChart({
      ...timeSeriesChart,
      candle: getConfigChartCandle(historicalIpcList, time)
    });
    
    setTimeout(() => {
      setShowGraph({
        area: false,
        candle: true
      });
    }, 0);
  };

  useEffect(() => {
    debugger;
    const configArea = getConfigChart(historicalIpcList);
    const configCandle = getConfigChartCandle(historicalIpcList, 15);

    configCandle.options.yaxis.min = minimumIpcValue.Precio;
    configCandle.options.yaxis.max = maximumIpcValue.Precio;

    setTimeSeriesChart({
      area: configArea,
      candle: configCandle
    });
  },[]);

  return (
    <section className='index--graph'>
      <div className='graph--header'>
        <div className='graph--date'>{moment(new Date(currentIpcValue.Fecha)).format('LL')}</div>
        <div className='graph--current_value'>
          <span className='current_text'>Valor Actual: </span>
          {new Intl.NumberFormat('de-DE').format(currentIpcValue.Precio)}
          <span className={`current_per ${symbolPercentage ? 'up' : 'down'}`}>
            {`(${symbolPercentage ? '+' : ''} ${currentIpcValue.Porcentaje}%)`}
          </span>
        </div>
        <div className='graph--options'>
          <figure className={`graph--area ${showGraph.area ? 'isActive' : ''}`} onClick={handleClickArea}>
            <img src={areaIcon} alt=""/>
          </figure>
          <figure className={`graph--candle ${showGraph.candle ? 'isActive' : ''}`} onClick={handleCliclCandle}>
            <img src={candleIcon} alt=""/>
          </figure>
          {showGraph.candle && (
            <div className='graph--candle_options'>
              <p id='5' className={`candle__min ${timeGraphCandle.five ? 'isActive' : ''}`} onClick={handleChangeTimeCandle}>5</p>
              <p id='15' className={`candle__min ${timeGraphCandle.fifteen ? 'isActive' : ''}`} onClick={handleChangeTimeCandle}>15</p>
              <p id='30' className={`candle__min ${timeGraphCandle.thirty ? 'isActive' : ''}`} onClick={handleChangeTimeCandle}>30</p>
              <p id='60' className={`candle__min ${timeGraphCandle.hour ? 'isActive' : ''}`} onClick={handleChangeTimeCandle}>60</p>
            </div>
          )}
        </div>
      </div>
      <div className='graph--history'>
        <div className='graph--timeseries'>
          {showGraph.area && (
            <Chart
              options={timeSeriesChart.area.options}
              series={timeSeriesChart.area.series}
              width='100%'
              height='100%'
            />
          )}

          {showGraph.candle && (
            <Chart
              options={timeSeriesChart.candle.options}
              series={timeSeriesChart.candle.series}
              width='100%'
              height='100%'
              type="candlestick"
            />
          )}
        </div>
        <div className='graph--candlestick' />
      </div>
    </section>
  );
};

const mapStateToProps = (state) => {
  return {
    historicalIpcList: state.historicalIpcList,
    currentIpcValue: state.currentIpcValue,
    maximumIpcValue: state.maximumIpcValue,
    minimumIpcValue: state.minimumIpcValue,
  };
};

export default connect(mapStateToProps, null)(IndexGraph);
