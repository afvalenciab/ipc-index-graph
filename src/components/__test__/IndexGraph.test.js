import React from 'react';
import { create } from 'react-test-renderer';
import IndexGraph from '../IndexGraph';
import ProviderMock from '../../__mocks__/ProviderMocks';

describe('IndexGraph testing', () => {
  test('Match Snapshot', () => {
    const indexGraph = create(
      <ProviderMock>
        <IndexGraph />
      </ProviderMock>
    );
    expect(indexGraph.toJSON()).toMatchSnapshot();
  });
});