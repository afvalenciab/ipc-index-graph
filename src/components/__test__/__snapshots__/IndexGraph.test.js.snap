// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexGraph testing Match Snapshot 1`] = `
<section
  className="index--graph"
>
  <div
    className="graph--header"
  >
    <div
      className="graph--date"
    >
      Invalid date
    </div>
    <div
      className="graph--current_value"
    >
      <span
        className="current_text"
      >
        Valor Actual: 
      </span>
      0
      <span
        className="current_per up"
      >
        (+ 0%)
      </span>
    </div>
    <div
      className="graph--options"
    >
      <figure
        className="graph--area isActive"
        onClick={[Function]}
      >
        <img
          alt=""
          src="test-file-stubs"
        />
      </figure>
      <figure
        className="graph--candle "
        onClick={[Function]}
      >
        <img
          alt=""
          src="test-file-stubs"
        />
      </figure>
    </div>
  </div>
  <div
    className="graph--history"
  >
    <div
      className="graph--timeseries"
    >
      <div
        height="100%"
        options={
          Object {
            "chart": Object {
              "animations": Object {
                "easing": "linear",
                "enable": true,
                "speed": 300,
              },
              "height": 350,
              "stacked": false,
              "toolbar": Object {
                "autoSelected": "zoom",
              },
              "type": "area",
            },
            "dataLabels": Object {
              "enabled": false,
            },
            "markers": Object {
              "size": 0,
            },
            "stroke": Object {
              "width": 1.5,
            },
            "title": Object {
              "align": "left",
              "text": "^MMX IPC Mexico",
            },
            "tooltip": Object {
              "shared": false,
            },
            "xaxis": Object {
              "convertedCatToNumeric": false,
              "labels": Object {
                "formatter": [Function],
              },
              "type": "datetime",
            },
            "yaxis": Object {
              "title": Object {
                "text": "Precio",
              },
            },
          }
        }
        series={
          Array [
            Object {
              "data": Array [],
              "name": "IPC Index",
            },
          ]
        }
        type="line"
        width="100%"
      />
    </div>
    <div
      className="graph--candlestick"
    />
  </div>
</section>
`;
